File: android/app/src/main/java/com/yagami073/qobuzmobilev2/RNMitsuhaVisualizerView.kt
package com.yagami073.qobuzmobilev2

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Color
import android.view.View
import kotlin.math.sin
import kotlin.math.PI

class RNMitsuhaVisualizerView(context: Context) : View(context) {
    
    private val numberOfBars = 40
    private var audioLevel: Float = 0f
    private val targetLevels = FloatArray(numberOfBars) { 0f }
    private val currentLevels = FloatArray(numberOfBars) { 0f }
    private var phase: Float = 0f
    
    // Customizable properties
    var barColorValue: Int = Color.rgb(29, 185, 84)
        set(value) {
            field = value
            paint.color = value
            invalidate()
        }
    
    var barWidthValue: Float = 3f
    var barSpacingValue: Float = 2f
    var sensitivityValue: Float = 1.5f
    var smoothnessValue: Float = 0.2f
    var waveStyleValue: String = "mitsuha"
    
    private val paint = Paint().apply {
        color = barColorValue
        style = Paint.Style.FILL
        isAntiAlias = true
        setShadowLayer(10f, 0f, 0f, barColorValue)
    }
    
    private val path = Path()
    private var animationRunnable: Runnable? = null
    
    init {
        setLayerType(LAYER_TYPE_SOFTWARE, null) // Enable shadow layer
        startAnimation()
    }
    
    private fun startAnimation() {
        animationRunnable = object : Runnable {
            override fun run() {
                updateAnimation()
                invalidate()
                postDelayed(this, 16) // ~60fps
            }
        }
        post(animationRunnable)
    }
    
    private fun updateAnimation() {
        // Update phase for wave animation
        phase += 0.05f
        if (phase > PI.toFloat() * 2) {
            phase = 0f
        }
        
        // Generate wave levels
        for (i in 0 until numberOfBars) {
            val waveValue = sin(phase + i * 0.3f) * 0.5f + 0.5f
            val randomFactor = (0.8f + Math.random().toFloat() * 0.4f)
            targetLevels[i] = waveValue * randomFactor * sensitivityValue
        }
        
        // Smooth interpolation
        for (i in 0 until numberOfBars) {
            currentLevels[i] += (targetLevels[i] - currentLevels[i]) * smoothnessValue
        }
    }
    
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        
        if (width <= 0 || height <= 0) return
        
        when (waveStyleValue) {
            "mitsuha" -> drawMitsuhaStyle(canvas)
            else -> drawBarStyle(canvas)
        }
    }
    
    private fun drawMitsuhaStyle(canvas: Canvas) {
        path.reset()
        
        val centerY = height / 2f
        val spacing = width.toFloat() / numberOfBars
        
        // Top wave
        var isFirst = true
        for (i in 0 until numberOfBars) {
            val x = i * spacing
            val amplitude = currentLevels[i] * (height / 3f)
            val y = centerY - amplitude
            
            if (isFirst) {
                path.moveTo(x, y)
                isFirst = false
            } else {
                path.lineTo(x, y)
            }
        }
        
        // Bottom wave (mirror)
        for (i in numberOfBars - 1 downTo 0) {
            val x = i * spacing
            val amplitude = currentLevels[i] * (height / 3f)
            val y = centerY + amplitude
            path.lineTo(x, y)
        }
        
        path.close()
        
        // Draw with transparency
        paint.alpha = 204 // 0.8 * 255
        canvas.drawPath(path, paint)
    }
    
    private fun drawBarStyle(canvas: Canvas) {
        val centerY = height / 2f
        val totalWidth = width - (barSpacingValue * (numberOfBars - 1))
        val calculatedBarWidth = totalWidth / numberOfBars
        val actualBarWidth = minOf(calculatedBarWidth, barWidthValue)
        
        for (i in 0 until numberOfBars) {
            val x = i * (actualBarWidth + barSpacingValue)
            val amplitude = currentLevels[i] * (height / 2f)
            
            val left = x
            val top = centerY - amplitude
            val right = x + actualBarWidth
            val bottom = centerY + amplitude
            
            canvas.drawRoundRect(
                left, top, right, bottom,
                actualBarWidth / 2, actualBarWidth / 2,
                paint
            )
        }
    }
    
    fun setAudioLevel(level: Float) {
        audioLevel = level
    }
    
    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        animationRunnable?.let { removeCallbacks(it) }
    }
}

File: ios/qobuzmobilev2/RNMitsuhaVisualizerView.swift
import UIKit
import AVFoundation

@objc(RNMitsuhaVisualizerView)
class RNMitsuhaVisualizerView: UIView {
  
  private var displayLink: CADisplayLink?
  private var waveLayers: [CAShapeLayer] = []
  private let numberOfBars: Int = 40
  private var audioLevel: Float = 0.0
  private var targetLevels: [CGFloat] = []
  private var currentLevels: [CGFloat] = []
  private var phase: CGFloat = 0.0
  
  // Customizable properties
  @objc var barColor: UIColor = UIColor(red: 29/255, green: 185/255, blue: 84/255, alpha: 1.0) {
    didSet {
      updateColors()
    }
  }
  
  @objc var barWidth: CGFloat = 3.0
  @objc var barSpacing: CGFloat = 2.0
  @objc var sensitivity: CGFloat = 1.5
  @objc var smoothness: CGFloat = 0.2
  @objc var waveStyle: String = "mitsuha" // "mitsuha" or "bar"
  
  override init(frame: CGRect) {
    super.init(frame: frame)
    setupView()
  }
  
  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setupView()
  }
  
  private func setupView() {
    backgroundColor = .clear
    clipsToBounds = true
    
    // Initialize levels
    targetLevels = Array(repeating: 0.0, count: numberOfBars)
    currentLevels = Array(repeating: 0.0, count: numberOfBars)
    
    // Setup display link for animation
    displayLink = CADisplayLink(target: self, selector: #selector(updateAnimation))
    displayLink?.add(to: .main, forMode: .common)
    displayLink?.preferredFramesPerSecond = 60
  }
  
  private func updateColors() {
    for layer in waveLayers {
      layer.fillColor = barColor.cgColor
      layer.strokeColor = barColor.cgColor
    }
  }
  
  @objc private func updateAnimation() {
    guard bounds.width > 0 && bounds.height > 0 else { return }
    
    // Update phase for wave animation
    phase += 0.05
    if phase > .pi * 2 {
      phase = 0
    }
    
    // Generate random-ish levels if no audio data (for demo)
    // In production, this should be driven by actual audio analysis
    for i in 0..<numberOfBars {
      let waveValue = sin(phase + CGFloat(i) * 0.3) * 0.5 + 0.5
      let randomFactor = CGFloat.random(in: 0.8...1.2)
      targetLevels[i] = waveValue * randomFactor * sensitivity
    }
    
    // Smooth interpolation
    for i in 0..<numberOfBars {
      currentLevels[i] += (targetLevels[i] - currentLevels[i]) * smoothness
    }
    
    // Draw based on style
    if waveStyle == "mitsuha" {
      drawMitsuhaStyle()
    } else {
      drawBarStyle()
    }
  }
  
  private func drawMitsuhaStyle() {
    // Clear previous layers
    waveLayers.forEach { $0.removeFromSuperlayer() }
    waveLayers.removeAll()
    
    let centerY = bounds.height / 2
    let totalWidth = bounds.width
    let spacing = totalWidth / CGFloat(numberOfBars)
    
    // Create path for wave
    let path = UIBezierPath()
    var isFirst = true
    
    // Top wave
    for i in 0..<numberOfBars {
      let x = CGFloat(i) * spacing
      let amplitude = currentLevels[i] * (bounds.height / 3)
      let y = centerY - amplitude
      
      if isFirst {
        path.move(to: CGPoint(x: x, y: y))
        isFirst = false
      } else {
        path.addLine(to: CGPoint(x: x, y: y))
      }
    }
    
    // Bottom wave (mirror)
    for i in (0..<numberOfBars).reversed() {
      let x = CGFloat(i) * spacing
      let amplitude = currentLevels[i] * (bounds.height / 3)
      let y = centerY + amplitude
      path.addLine(to: CGPoint(x: x, y: y))
    }
    
    path.close()
    
    // Create gradient layer
    let shapeLayer = CAShapeLayer()
    shapeLayer.path = path.cgPath
    shapeLayer.fillColor = barColor.cgColor
    shapeLayer.opacity = 0.8
    
    // Add glow effect
    shapeLayer.shadowColor = barColor.cgColor
    shapeLayer.shadowOffset = .zero
    shapeLayer.shadowRadius = 10
    shapeLayer.shadowOpacity = 0.6
    
    layer.addSublayer(shapeLayer)
    waveLayers.append(shapeLayer)
  }
  
  private func drawBarStyle() {
    // Clear previous layers
    waveLayers.forEach { $0.removeFromSuperlayer() }
    waveLayers.removeAll()
    
    let centerY = bounds.height / 2
    let totalWidth = bounds.width - (barSpacing * CGFloat(numberOfBars - 1))
    let calculatedBarWidth = totalWidth / CGFloat(numberOfBars)
    let actualBarWidth = min(calculatedBarWidth, barWidth)
    
    for i in 0..<numberOfBars {
      let x = CGFloat(i) * (actualBarWidth + barSpacing)
      let amplitude = currentLevels[i] * (bounds.height / 2)
      
      // Create bar
      let barLayer = CAShapeLayer()
      let barPath = UIBezierPath(roundedRect: CGRect(
        x: x,
        y: centerY - amplitude,
        width: actualBarWidth,
        height: amplitude * 2
      ), cornerRadius: actualBarWidth / 2)
      
      barLayer.path = barPath.cgPath
      barLayer.fillColor = barColor.cgColor
      
      // Add glow
      barLayer.shadowColor = barColor.cgColor
      barLayer.shadowOffset = .zero
      barLayer.shadowRadius = 5
      barLayer.shadowOpacity = 0.5
      
      layer.addSublayer(barLayer)
      waveLayers.append(barLayer)
    }
  }
  
  @objc func setAudioLevel(_ level: NSNumber) {
    audioLevel = level.floatValue
  }
  
  override func layoutSubviews() {
    super.layoutSubviews()
    // Redraw when bounds change
  }
  
  deinit {
    displayLink?.invalidate()
    displayLink = nil
  }
}
